File: app/src/main/java/com/kinvey/sample/kitchensink/appData/GetFragment.java
/** 
 * Copyright (c) 2014 Kinvey Inc.
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 * 
 * http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations under
 * the License.
 * 
 */
package com.kinvey.sample.kitchensink.appData;

import android.os.Bundle;
import android.view.View;
import android.widget.*;

import com.kinvey.android.callback.KinveyReadCallback;
import com.kinvey.android.store.DataStore;
import com.kinvey.java.core.KinveyClientCallback;
import com.kinvey.java.model.KinveyReadResponse;
import com.kinvey.java.store.StoreType;
import com.kinvey.sample.kitchensink.*;
import com.kinvey.sample.kitchensink.R;

import org.jetbrains.annotations.Nullable;

import java.util.List;

/**
 * @author edwardf
 * @since 2.0
 */
public class GetFragment extends UseCaseFragment implements View.OnClickListener {

    private TextView currentCountText;
    private Spinner viewingIDSpinner;
    private TextView currentNameText;
    private TextView currentIDText;
    private Button getItBtn;

    private DataStore<MyEntity> entityStore = null;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        entityStore = DataStore.collection(KitchenSinkActivity.collectionName, MyEntity.class, StoreType.AUTO, getClient());
    }

    @Override
    public void onResume() {
        super.onResume();
        getCount();
    }

    @Override
    public int getViewID() {
        return R.layout.feature_appdata_get;
    }

    @Override
    public void bindViews(View v) {
        currentCountText = (TextView) v.findViewById(R.id.appdata_get_current_count);
        viewingIDSpinner = (Spinner) v.findViewById(R.id.appdata_get_id_spinner);
        currentNameText = (TextView) v.findViewById(R.id.appdata_get_name_value);
        currentIDText = (TextView) v.findViewById(R.id.appdata_get_id_value);
        getItBtn = (Button) v.findViewById(R.id.appdata_get_button);
        getItBtn.setOnClickListener(this);
        viewingIDSpinner.setAdapter(new ArrayAdapter<String>(getActivity(), android.R.layout.simple_spinner_dropdown_item, new String[]{"--"}));
        viewingIDSpinner.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
            @Override
            public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
                if (viewingIDSpinner.getSelectedItem().toString().equals("--")){
                    return;
                }
                findSelectedItem(viewingIDSpinner.getSelectedItem().toString());
            }
            @Override
            public void onNothingSelected(AdapterView<?> parent) {
                //To change body of implemented methods use File | Settings | File Templates.
            }
        });
    }

    @Override
    public String getTitle() {
        return "Get";
    }

    public void getIt() {
        getCount();
        if (viewingIDSpinner.getSelectedItem() == null) {
            return;
        }
        findSelectedItem(viewingIDSpinner.getSelectedItem().toString());
    }

    private void findSelectedItem(String id) {
        entityStore.find(id, new KinveyClientCallback<MyEntity>() {
            @Override
            public void onSuccess(MyEntity result) {
                currentNameText.setText(result.getName());
                currentIDText.setText(result.getId());
            }
            @Override
            public void onFailure(Throwable error) {
                AndroidUtil.toast(GetFragment.this, "something went wrong on getEntity ->" + error.getMessage());
            }
        });
    }

    @Override
    public void onClick(View v) {
        if (v == getItBtn) { getIt(); }
    }

    private void updateSpinner(List<MyEntity> result) {
        if (getActivity() == null) { return; }
        String[] ids = new String[result.size()];
        int count = 0;
        for (MyEntity item : result) {
            ids[count] = item.getId();
            count++;
        }
        viewingIDSpinner.setAdapter(new ArrayAdapter(getActivity(), android.R.layout.simple_spinner_dropdown_item, ids));
    }

    private void getCount() {
        entityStore.find(new KinveyReadCallback<MyEntity>() {
            @Override
            public void onSuccess(@Nullable KinveyReadResponse<MyEntity> result) {
                currentCountText.setText(String.valueOf(result.getResult().size()));
                updateSpinner(result.getResult());
            }
            @Override
            public void onFailure(Throwable error) {
                AndroidUtil.toast(getActivity(), "something went wrong on get->" + error.getMessage());
            }
        });
    }
}

File: app/src/main/java/com/kinvey/sample/kitchensink/appData/GetFragment.kt
/**
 * Copyright (c) 2019 Kinvey Inc.
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations under
 * the License.
 *
 */
package com.kinvey.sample.kitchensink.appData

import android.content.Context
import android.os.Bundle
import android.view.View
import android.view.View.OnClickListener
import android.widget.*
import android.widget.AdapterView.OnItemSelectedListener
import com.kinvey.android.callback.KinveyReadCallback
import com.kinvey.android.store.DataStore
import com.kinvey.java.core.KinveyClientCallback
import com.kinvey.java.model.KinveyReadResponse
import com.kinvey.java.store.StoreType
import com.kinvey.sample.kitchensink.*
import kotlinx.android.synthetic.main.feature_appdata_get.*

/**
 * @author edwardf
 * @since 2.0
 */
class GetFragment : UseCaseFragment(), OnClickListener {

    private var entityStore: DataStore<MyEntity>? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        entityStore = DataStore.collection(Constants.COLLECTION_NAME, MyEntity::class.java, StoreType.AUTO, client)
    }

    override fun onResume() {
        super.onResume()
        count()
    }

    override val viewID = R.layout.feature_appdata_get

    override val title = "Get"

    override fun initViews(v: View) {
        getBtn?.setOnClickListener(this)
        getIdSpinner?.adapter = ArrayAdapter(activity!!, android.R.layout.simple_spinner_dropdown_item, arrayOf("--"))
        getIdSpinner?.onItemSelectedListener = object : OnItemSelectedListener {
            override fun onItemSelected(parent: AdapterView<*>?, view: View?, position: Int, id: Long) {
                if (getIdSpinner?.selectedItem.toString() == "--") {
                    return
                }
                findSelectedItem(getIdSpinner?.selectedItem.toString())
            }
            override fun onNothingSelected(parent: AdapterView<*>?) {
                //To change body of implemented methods use File | Settings | File Templates.
            }
        }
    }

    private fun getIt() {
        count()
        if (getIdSpinner?.selectedItem == null) {
            return
        }
        findSelectedItem(getIdSpinner?.selectedItem.toString())
    }

    private fun findSelectedItem(id: String) {
        entityStore?.find(id, object : KinveyClientCallback<MyEntity> {
            override fun onSuccess(result: MyEntity) {
                getNameValueText?.text = result.name
                getIdValueText?.text = result.id
            }
            override fun onFailure(error: Throwable) {
                AndroidUtil.toast(this@GetFragment, "something went wrong on getEntity -> ${error.message}")
            }
        })
    }

    override fun onClick(v: View) {
        if (v === getBtn) { getIt() }
    }

    private fun updateSpinner(result: List<MyEntity>?) {
        if (activity == null) { return }
        val ids = result?.map { item -> item.id }?.toTypedArray() ?: arrayOf()
        getIdSpinner?.adapter = ArrayAdapter<String>(context as Context, android.R.layout.simple_spinner_dropdown_item, ids)
    }

    private fun count() {
        entityStore?.find(object : KinveyReadCallback<MyEntity> {
            override fun onSuccess(result: KinveyReadResponse<MyEntity>?) {
                getCurrentCountText?.text = result?.result?.size.toString()
                updateSpinner(result?.result)
            }
            override fun onFailure(error: Throwable) {
                AndroidUtil.toast(activity, "something went wrong on get-> ${error.message}")
            }
        })
    }
}